src/xfieldhistory.h

```diff
@@ -0,0 +1,244 @@
+// Copyright (c) 2019-2022 Chaintope Inc.
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef TAPYRUS_XFIELDHISTORY_H
+#define TAPYRUS_XFIELDHISTORY_H
+
+#include <policy/policy.h>
+#include <federationparams.h>
+/* 
+ * struct to store xfieldValue, block hash and height for every xfield update in the blockchain.
+ */
+
+struct XFieldChange {
+    XFieldData xfieldValue;
+    uint32_t height;
+    uint256 blockHash;
+
+    XFieldChange():xfieldValue(), height(0), blockHash(){}
+    explicit XFieldChange(const XFieldData& in_data, uint32_t in_height, uint256 in_blockHash):xfieldValue(in_data), height(in_height), blockHash(in_blockHash){}
+    explicit XFieldChange(const XFieldData&& in_data, uint32_t in_height, uint256 in_blockHash):xfieldValue(in_data), height(in_height), blockHash(in_blockHash){}
+    XFieldChange(const XFieldChange& copy):xfieldValue(copy.xfieldValue),height(copy.height),blockHash(copy.blockHash){}
+    ~XFieldChange() = default;
+
+    bool operator==(const XFieldChange& copy) const {
+        return xfieldValue == copy.xfieldValue
+            && height == copy.height
+            && blockHash == copy.blockHash;
+    }
+
+    template<typename Stream>
+    void Serialize(Stream& s) const {
+        switch(GetXFieldTypeFrom(xfieldValue))
+        {
+            case TAPYRUS_XFIELDTYPES::AGGPUBKEY:
+                ::Serialize(s, boost::get<XFieldAggPubKey>(xfieldValue)); break;
+            case TAPYRUS_XFIELDTYPES::MAXBLOCKSIZE:
+                ::Serialize(s, boost::get<XFieldMaxBlockSize>(xfieldValue)); break;
+            case TAPYRUS_XFIELDTYPES::NONE:
+            default:
```

Comment:
It may not be necessary here. serialise is always invoked from within block or block index. The RPC used to create block takes care of. incorrect xfield type. So we can expect all the code reading here to have the correct  xfield type.